### PR TITLE
Ensure experimental settings run against experimental features

### PR DESCRIPTION
**Pull request by jaredpar**
Whlile working on another change I noticed that our testing support for experimental features didn't actually verify the underlying feature was experimental.  Hence helpers like ParseExperimental(...) would continue to function without error when tuples / ref returns were attached to the /langver switch.  This creates maintenance issues in our test framework as it's unclear what we're actually testing.

I refactored the code involved to verify the features are actually experimental and removed some troublesome helpers.

### CODE REVIEW

**Comment by jaredpar** [comment]
CC @dotnet/roslyn

**Comment by jaredpar** [comment]
CC @dotnet/roslyn-compiler

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb [hunk 1]
```diff
@@ -32,17 +28,26 @@ End Class
...
+    Friend Function WithExperimental(options As VisualBasicParseOptions, ParamArray features As Feature()) As VisualBasicParseOptions
+        If features.Length = 0 Then
+            Throw New InvalidOperationException("Need at least one feature to enable")
+        End If
+
+        Dim list As New List(Of KeyValuePair(Of String, String))
+        For Each feature In features
+            Dim flag = feature.GetFeatureFlag()
+            If flag Is Nothing Then
+                Throw New InvalidOperationException($"{feature} is not an experimental featrure")
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Typo "featrure"?

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb [hunk 2]
```diff
@@ -32,17 +28,26 @@ End Class
...
-        Return options.WithFeature("Deterministic", "true")
+    Friend Function WithExperimental(options As VisualBasicParseOptions, ParamArray features As Feature()) As VisualBasicParseOptions
+        If features.Length = 0 Then
+            Throw New InvalidOperationException("Need at least one feature to enable")
+        End If
+
+        Dim list As New List(Of KeyValuePair(Of String, String))
+        For Each feature In features
+            Dim flag = feature.GetFeatureFlag()
+            If flag Is Nothing Then
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Should we be checking for a specific flag here?

Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Or is this a feature id? I guess the name is confusing somewhat.

---
In reply to: [67381960](https://github.com/dotnet/roslyn/pull/12039#discussion_r67381960) [](ancestors = 67381960)

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs [hunk 3]
```diff
@@ -44,34 +41,52 @@ public static class TestOptions
...
         public static CSharpParseOptions WithRefsFeature(this CSharpParseOptions options)
         {
-            return options.WithFeature(MessageID.IDS_FeatureRefLocalsReturns.RequiredFeature(), "true");
+            return WithExperimental(options, MessageID.IDS_FeatureRefLocalsReturns);
         }
 
         public static CSharpParseOptions WithTuplesFeature(this CSharpParseOptions options)
         {
-            return options.WithFeature("tuples", "true");
+            return WithExperimental(options, MessageID.IDS_FeatureTuples);
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: I think we have another file that has similar helpers. Should we make similar changes there?

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb (hunk 1 above)
Comment (jaredpar) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: Fixed.

File: src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb (hunk 2 above)
Comment (jaredpar) [inline src/Compilers/Test/Utilities/VisualBasic/TestOptions.vb]: This is the name of the flag that is passed to the `/feature:` command line option.  I renamed the variable to `flagName` to try and make it clearer.

File: src/Compilers/Test/Utilities/CSharp/TestOptions.cs (hunk 3 above)
Comment (jaredpar) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: I did a search for all uses of `WithFeatures` in the code base.  While I didn't find any other file with a helper set I did find some more incorrect / loose uses that need to be cleaned up.  I'll migrate them all over to the helpers I defined.

Comment (jaredpar) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: Hmm, It looks like we do have a separate helper file that is virtually identical to this one ... weird.  Looking into that.

Comment (jaredpar) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: That file was a needless copy.  I just deleted it, added the one I fixed up as a linked file.

Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp/TestOptions.cs]: I don't see any changes to project files.

**Comment by jaredpar** [comment]
Updated to respond to feedback.

File: src/Compilers/Test/Utilities/CSharp.Desktop/CSharpTestBase.cs [hunk 4]
```diff
@@ -405,8 +406,9 @@ public static SyntaxTree ParseWithRoundTripCheck(string text, CSharpParseOptions
                 assemblyName: assemblyName);
         }
 
-        public static CSharpCompilation CreateExperimentalCompilationWithMscorlib45(
+        internal static CSharpCompilation CreateExperimentalCompilationWithMscorlib45(
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp.Desktop/CSharpTestBase.cs]: Is this file also a dupe of the one in CSharp folder?

File: src/Compilers/Test/Utilities/CSharp.Desktop/SemanticModelTestBase.cs [hunk 5]
```diff
@@ -176,9 +176,9 @@ private static ConstructorInitializerSyntax GetFirstConstructorInitializer(Synta
             return GetSemanticInfoForTest<TNode>(compilation);
         }
 
-        protected CompilationUtils.SemanticInfoSummary GetSemanticInfoForTestExperimental<TNode>(string testSrc) where TNode : SyntaxNode
+        internal CompilationUtils.SemanticInfoSummary GetSemanticInfoForTestExperimental<TNode>(string testSrc, MessageID feature) where TNode : SyntaxNode
```
Comment (AlekseyTs) [inline src/Compilers/Test/Utilities/CSharp.Desktop/SemanticModelTestBase.cs]: Is this file also a dupe of the one in CSharp folder?

**Comment by AlekseyTs** [comment]
LGTM. Though it looks like we might have two more files duplicated in Test/Utilities/CSharp.Desktop.

**Pull request merged**